### PR TITLE
[Graphbolt] Rewrite torch::multinomial to improve sampling performance

### PR DESCRIPTION
**Pull request by RamonZhou**
## Description
The non-uniform Neighbor sampling didn't gain much boost in #6152. Here I rewrite this branch according to the implementation in ATEN.
The correctness of this random sampling process still needs to be evaluated.

References:
- https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/Distributions.cpp
- https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cpu/MultinomialKernel.cpp

## Performance

Tested performance on 6 different graphs with settings:
Graph | Num of nodes | Num of neighbors per etype | Node types | Edge types | Fanouts
:-:|:-:|:-:|:-:|:-:|:-
G1 | 10,000,000 | 5 | 3 | 2 | [3,4]
G2 | 100 | 1,000,000 | 3 | 2 | [899999, 911111]
G3 | 10,000 | 100,000 | 3 | 2 | [6000, 9000]
G4 | 10,000 | 5,000 | 30 | 20 | [1000, 8000, ..., 3000]
G5 (homo) | 1,000,000 | 100 | 1 | 1 | [80]
G6 | 10,000 | 100,000 | 5 | 4 | [100, 200, 300, 400]

Bar charts:

- `v0`: Old version before #6152.
- `v1`: The version after #6152.
- `v2`: This PR.

V2-NF
V2-NT

`replace` | Version | G1 | G2 | G3 | G4 | G5 | G6
:-:|:-:|:-:|:-:|:-:|:-:|:-:|:-:
False | v2 | 29590101.2 | 6913956.7 | 4199777.8 | 2408906.5 | 1728034.8 | 6894497.9
False | v1 | 29084539.2 | 16515189.3 | 38836236.9 | 17375014.6 | 4801887.6 | 78171954.6
False | v0 | 38688025.1 | 18178196.6 | 49979084.7 | 26732285.2 | 5280066.0 | 89704824.2
False | v2 > v0 | +30.7% | +162.9% | **+1090.0%** | **+1009.7%** | +205.6% | **+1201.1%**
False | v1 > v0 | +33.0% | +10.1% | +28.7% | +53.9% | +10.0% | +14.8%
True | v2 | 30968475.6 | 25552874.8 | 2487178.8 | 3232033.1 | 1798778.5 | 2356059.0
True | v1 | 45981538.9 | 35921067.0 | 24986967.4 | 44202859.5 | 9608167.8 | 10192462.6
True | v0 | 54829759.8 | 40574989.3 | 35469360.9 | 53122429.3 | 10416474.4 | 13245708.9
True | v2 > v0 | +77.1% | +58.8% | **+1326.1%** | **+1543.6%** | **+479.1%** | **+462.2%**
True | v1 > v0 | +19.2% | +13.0% | +42.0% | +20.2% | +8.4% | +30.0%

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Changes
- Rewrote `torch::multinomial`.

### CODE REVIEW

**Comment by mfbalin** [comment]
@RamonZhou How is the correctness going to be evaluated? If there is going to be a test, can we evaluate whether Labor works correctly as well?

**Comment by RamonZhou** [comment]
@mfbalin I think it's not easy to evaluate. My plan is to extract the sampling logic and run it by a large amount of times then test the sampled distribution.

File: graphbolt/src/csc_sampling_graph.cc [hunk 1]
```diff
@@ -636,11 +636,108 @@ inline int64_t NonUniformPick(
...
-        fanout * sizeof(PickedType));
+    if (fanout == 0) return 0;
+    TORCH_CHECK(
+        ((local_probs.max() < INFINITY) & (local_probs.min() >= 0))
+            .item()
+            .to<bool>(),
+        "Invalid probs_or_mask (contains either `inf`, `nan` or element < 0).");
+    TORCH_CHECK(
+        !(local_probs.sum() == 0).item().to<bool>(),
+        "Invalid probs_or_mask (sum of probabilities <= 0).");
```
Comment (frozenbugs) [inline graphbolt/src/csc_sampling_graph.cc:647]: In previous check, there is local_probs.min() >= 0. Why do we need to check !local_probs.sum() == 0?

File: graphbolt/src/csc_sampling_graph.cc [hunk 2]
```diff
@@ -636,11 +636,108 @@ inline int64_t NonUniformPick(
...
+            // of argmax or topk. Then we have s = argmax( p / (-log(eps)) )
+            // where eps ~ U(0, 1). We can also simplify the formula above by s
+            // = argmax( p / q ) where q ~ Exp(1)
+            std::exponential_distribution<> exp_distrib(1);
+            if (fanout == 1) {
+              // Return argmax(p / q).
+              scalar_t max_prob = 0;
+              PickedType max_prob_index = -1;
+              for (auto i = 0; i < num_positive_probs; ++i) {
+                scalar_t cur_prob =
```
Comment (frozenbugs) [inline graphbolt/src/csc_sampling_graph.cc:669]: what's cur_prob?

File: graphbolt/src/csc_sampling_graph.cc [hunk 3]
```diff
@@ -636,11 +636,108 @@ inline int64_t NonUniformPick(
...
+            // Calculate cumulative sum of probabilities.
+            std::vector<scalar_t> cum_probs(num_positive_probs);
+            scalar_t sum_probs = 0;
+            for (auto i = 0; i < num_positive_probs; ++i) {
+              sum_probs += local_probs_data_ptr[positive_probs_indices_ptr[i]];
+              cum_probs[i] = sum_probs;
+            }
+            // Normalize.
+            if ((sum_probs > 0) ||
+                ((sum_probs < 1.00001) && (sum_probs > 0.99999))) {
```
Comment (frozenbugs) [inline graphbolt/src/csc_sampling_graph.cc:714]: The set of (sum_probs < 1.00001) && (sum_probs > 0.99999) is already included in the set of (sum_probs > 0), why we need ||?

**Comment by peizhou001** [comment]
Is there a separate single test against` torch::multinomia`l and the new algorithm?

File: graphbolt/src/csc_sampling_graph.cc [hunk 4]
```diff
@@ -636,11 +636,106 @@ inline int64_t NonUniformPick(
     return num_positive_probs;
   } else {
     if (!replace) fanout = std::min(fanout, num_positive_probs);
-    std::memcpy(
-        picked_data_ptr,
-        (torch::multinomial(local_probs, fanout, replace) + offset)
-            .data_ptr<PickedType>(),
-        fanout * sizeof(PickedType));
+    if (fanout == 0) return 0;
+    TORCH_CHECK(
```
Comment (peizhou001) [inline graphbolt/src/csc_sampling_graph.cc:640]: Avoid this expensive check, we can do it offline.

File: graphbolt/src/csc_sampling_graph.cc [hunk 5]
```diff
@@ -636,11 +636,106 @@ inline int64_t NonUniformPick(
...
+            .to<bool>(),
+        "Invalid probs_or_mask (contains either `inf`, `nan` or element < 0).");
+    AT_DISPATCH_ALL_TYPES(
+        local_probs.scalar_type(), "MultinomialSampling", ([&] {
+          auto local_probs_data_ptr = local_probs.data_ptr<scalar_t>();
+          auto positive_probs_indices_ptr =
+              positive_probs_indices.data_ptr<PickedType>();
+
+          std::random_device rd;
+          std::mt19937 gen(rd());
```
Comment (peizhou001) [inline graphbolt/src/csc_sampling_graph.cc:652]: Put this to a separate class and use pcg as random seed.

File: graphbolt/src/csc_sampling_graph.cc [hunk 6]
```diff
@@ -636,11 +636,106 @@ inline int64_t NonUniformPick(
...
+            }
+            std::uniform_real_distribution<> uniform(0, 1);
+            for (auto i = 0; i < fanout; ++i) {
+              // Sample a probability mass from a uniform distribution.
+              double uniform_sample = uniform(gen);
+              // Use a binary search to find the index.
+              int left_pointer = 0;
+              int right_pointer = num_positive_probs;
+              int mid_pointer;
+              while (right_pointer - left_pointer > 0) {
```
Comment (peizhou001) [inline graphbolt/src/csc_sampling_graph.cc:725]: Is there any std function can do this binary search?

File: graphbolt/src/csc_sampling_graph.cc [hunk 7]
```diff
@@ -636,11 +636,106 @@ inline int64_t NonUniformPick(
...
+                std::nth_element(
+                    q.begin(), q.begin() + fanout - 1, q.end(), std::greater{});
+                for (auto i = 0; i < fanout; ++i) {
+                  picked_data_ptr[i] = q[i].second + offset;
+                }
+              }
+            }
+          } else {
+            // Calculate cumulative sum of probabilities.
+            std::vector<scalar_t> cum_probs(num_positive_probs);
```
Comment (peizhou001) [inline graphbolt/src/csc_sampling_graph.cc:705]: `perfix_sum_probs`

File: graphbolt/src/csc_sampling_graph.cc [hunk 8]
```diff
@@ -636,11 +636,106 @@ inline int64_t NonUniformPick(
...
+        "Invalid probs_or_mask (contains either `inf`, `nan` or element < 0).");
+    AT_DISPATCH_ALL_TYPES(
+        local_probs.scalar_type(), "MultinomialSampling", ([&] {
+          auto local_probs_data_ptr = local_probs.data_ptr<scalar_t>();
+          auto positive_probs_indices_ptr =
+              positive_probs_indices.data_ptr<PickedType>();
+
+          std::random_device rd;
+          std::mt19937 gen(rd());
+          if (!replace) {
```
Comment (peizhou001) [inline graphbolt/src/csc_sampling_graph.cc:653]: Do we need without replacement implementation?

File: graphbolt/src/csc_sampling_graph.cc [hunk 9]
```diff
@@ -636,11 +636,106 @@ inline int64_t NonUniformPick(
...
+              sum_probs += local_probs_data_ptr[positive_probs_indices_ptr[i]];
+              cum_probs[i] = sum_probs;
+            }
+            // Normalize.
+            if ((sum_probs < 1.00001) && (sum_probs > 0.99999)) {
+              for (auto i = 0; i < num_positive_probs; ++i) {
+                cum_probs[i] /= sum_probs;
+              }
+            }
+            std::uniform_real_distribution<> uniform(0, 1);
```
Comment (peizhou001) [inline graphbolt/src/csc_sampling_graph.cc:717]: Move all random related code to Random Engine.

File: graphbolt/src/csc_sampling_graph.cc [hunk 10]
```diff
@@ -636,11 +636,106 @@ inline int64_t NonUniformPick(
...
-        (torch::multinomial(local_probs, fanout, replace) + offset)
-            .data_ptr<PickedType>(),
-        fanout * sizeof(PickedType));
+    if (fanout == 0) return 0;
+    TORCH_CHECK(
+        ((local_probs.max() < INFINITY) & (local_probs.min() >= 0))
+            .item()
+            .to<bool>(),
+        "Invalid probs_or_mask (contains either `inf`, `nan` or element < 0).");
+    AT_DISPATCH_ALL_TYPES(
```
Comment (peizhou001) [inline graphbolt/src/csc_sampling_graph.cc:645]: Make the dispatch more specific.

**Comment by RamonZhou** [comment]
@peizhou001 I've run some tests locally and the it seems to be correct (the output distributions of these two are the same).
Plus, this is not actually a rewriting of algorithm. All the checks, if-conditions, algorithms and the binary search are the same as the internal implementation of `ATen`. I only did some simplification and removed some useless instructions.

File: graphbolt/src/csc_sampling_graph.cc (hunk 3 above)
Comment (RamonZhou) [reply graphbolt/src/csc_sampling_graph.cc]: I actually copied and pasted this from ATen. I'm now a little confused. I don't think it is what their developers intended to write LOL.

**Comment by peizhou001** [comment]
> @peizhou001 I've run some tests locally and the it seems to be correct (the output distributions of these two are the same). Plus, this is not actually a rewriting of algorithm. All the checks, if-conditions, algorithms and the binary search are the same as the internal implementation of `ATen`. I only did some simplification and removed some useless instructions.

Got it, but also we first need to know if the slowdown comes from `memcpy` or not.

**Comment by RamonZhou** [comment]
@peizhou001 I did a simple test. I combined different fanouts and replace settings together and measured the total time.
```cpp
n = 1000000;
for (int i = 1; i <= 1000; ++ i) {
    fanout = i * 1000;
    replace = (i & 1) > 0;
    Sample(0, n, fanout, replace, ...); // Replace Sample function with different implementations.
}
```

Results are:
- `torch::multinomial`: 96573.2ms
- `std::memcpy(torch::multinomial)`: 98845.2ms
- My code: 74089.2ms

So I think the gain not only comes from the elimination of `std::memcpy`.

File: graphbolt/src/csc_sampling_graph.cc (hunk 2 above)
Comment (RamonZhou) [reply graphbolt/src/csc_sampling_graph.cc:668]: Added a comment for this.

File: graphbolt/src/random.h [hunk 11]
```diff
@@ -69,6 +69,25 @@ class RandomEngine {
     return dist(rng_);
   }
 
+  /**
+   * @brief Generate a uniform random real number in [low, high).
+   */
+  template <typename T>
+  T RandomUniform(T lower, T upper) {
```
Comment (peizhou001) [inline graphbolt/src/random.h:76]: Remove `Random` in the name as this is already denoted in the class name.

File: graphbolt/src/random.h [hunk 12]
```diff
@@ -69,6 +69,25 @@ class RandomEngine {
...
+    std::uniform_real_distribution<T> dist(lower, upper);
+    return dist(rng_);
+  }
+
+  /**
+   * @brief Generate random non-negative floating-point values according to
+   * exponential distribution. Probability density function: P(x|λ) = λe^(-λx)
+   */
+  template <typename T>
+  T RandomExponential(T lambda) {
```
Comment (peizhou001) [inline graphbolt/src/random.h:86]: Ditto.

**Review by peizhou001 — APPROVED** [review]

File: graphbolt/src/csc_sampling_graph.cc [hunk 13]
```diff
@@ -690,11 +696,90 @@ inline int64_t NonUniformPick(
...
+          auto positive_probs_indices_ptr =
+              positive_probs_indices.data_ptr<PickedType>();
+
+          if (!replace) {
+            // The algorithm is from gumbel softmax.
+            // s = argmax( logp - log(-log(eps)) ) where eps ~ U(0, 1)
+            // Here we can apply exp to the formula which will not affect result
+            // of argmax or topk. Then we have s = argmax( p / (-log(eps)) )
+            // where eps ~ U(0, 1). We can also simplify the formula above by s
+            // = argmax( p / q ) where q ~ Exp(1)
```
Comment (frozenbugs) [inline graphbolt/src/csc_sampling_graph.cc:712]: super nit:
```suggestion
            // where eps ~ U(0, 1). We can also simplify the formula above by
            // s = argmax( p / q ) where q ~ Exp(1)
```

File: graphbolt/src/csc_sampling_graph.cc [hunk 14]
```diff
@@ -690,11 +696,90 @@ inline int64_t NonUniformPick(
...
+            if (fanout == 1) {
+              // Return argmax(p / q).
+              scalar_t max_prob = 0;
+              PickedType max_prob_index = -1;
+              // We only care about the neighbors with non-zero probability.
+              for (auto i = 0; i < num_positive_probs; ++i) {
+                // Calculate (p / q) for the current neighbor.
+                scalar_t cur_prob =
```
Comment (frozenbugs) [inline graphbolt/src/csc_sampling_graph.cc:720]: What about just current_prob?
```suggestion
                scalar_t current_prob =
```

**Pull request merged**